### ISSUE TITLE
pam_systemd_home.so breaks password configuration

### ISSUE THREAD

**Issue by schaarsc** [systemd/systemd#30969]
### systemd version the issue has been seen with

254

### Used distribution

Debian 12

### CPU architectures issue was seen on

None

### Component

pam_systemd_home

### Expected behaviour you didn't see

permit passwd to change root password

### Unexpected behaviour you saw

error message: op=PAM:chouthtok res=failed

### Steps to reproduce the problem

Debian creates the following /etc/pam.d/common-password 
```
password        [success=2 default=ignore]      pam_systemd_home.so
password        [success=1 default=ignore]      pam_unix.so obscure use_authtok try_first_pass yescrypt

```
login as root
try to change password with
`passwd`

if pam_systemd_home is deactivated with pam-auth-update everything works as expected.

also if the pam_pwquality is in place the error does not occur
```
password        requisite                       pam_pwquality.so retry=3
password        [success=2 default=ignore]      pam_systemd_home.so
password        [success=1 default=ignore]      pam_unix.so obscure use_authtok try_first_pass yescrypt
```
this configuration works

**Comment by poettering** [comment]
what's the error message? what's in the logs?

**Comment by mbiebl** [comment]
A related bug report might be
https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1056166
which has more logs

**Comment by poettering** [comment]
hmm, i am a bit puzzled by your pam fragment. It's not the pam fragment we recommend, see pam_systemd(5)

**Comment by mbiebl** [comment]
Well, PAM is a mess and highly distro specific.
Which bit in particular do you see as problematic?

**Comment by schaarsc** [comment]
if I understand https://www.man7.org/linux/man-pages/man5/pam.d.5.html correctly, then [success= default=ignore] should be equivalent to sufficient.
I've also changed `password [success=2 default=ignore]  pam_systemd_home.so `
to `password sufficient  pam_systemd_home.so `
but the result is the same
```
# passwd
passwd: Authentication token manipulation error
passwd: password unchanged
```
the strange thing here is that if pam_systemd_home.so is listed first, then there is no prompt asking for a password

as for the log-files, unfortunately I cannot provide more information than what has already been provided in https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1056166

**Comment by poettering** [comment]
So, this is not a systemd issue afaics. You configure pam_unix with "use_authtok". The way I read the pam_unix sources this has the effect that the module will refuse to ask for passwords on its own but instead insist that a previous module in the stack already acquired one and inserted it into the pam context. It will never ask the user directly.

This is supposed to be used to gether with some pw strength checking module plugged in front of it, i.e. cracklib, pwq or so, which asks for the pw, checks it, and then leaves it in the stack context for pam_unix.

See pam_get_authtok(3) man page (the part about "OPTIONS" where it describes what happens if you use "use_authtok". pam_unix uses pam_get_authtok() to query the pw, and thus will fail if nothing is cached.

pam_systemd_home will not do a thing for users it doesn't manage on its own, hence it has no effect on any of this at all.

So you basically have two options:

1. add craclick/pwq or something like that to your pam stack, preceeding pam_unix.
2. drop "use_authtok" from the switches.

but systemd won't change any of this.

Closing.

**Comment by poettering** [comment]
this is wrong btw:

```
password        requisite                       pam_pwquality.so retry=3
password        [success=2 default=ignore]      pam_systemd_home.so
password        [success=1 default=ignore]      pam_unix.so obscure use_authtok try_first_pass yescrypt
```

the first lines should be switched: pam_systemd_home.so will ask on its own for passwords, and check them against pwq. This should not be done via the pam_pwq before.

(of course if you swap the lines you need to adjust the success= jump target)

**Comment by poettering** [comment]
our dumbed down example PAM stacks use use_authtok too btw even though we do not include pwq in it. I prepped a fix for that in #30992
